packages/flutter/lib/src/rendering/layer.dart

```diff
@@ -41,6 +41,33 @@ abstract class Layer extends AbstractNode with DiagnosticableTreeMixin {
   @override
   ContainerLayer get parent => super.parent;
 
+  // Whether this layer has any changes since it's last added to the engine
+  // using [addToScene]. Initialized to true as a new layer has never called
+  // [addToScene].
+  bool _isDirty = true;
+
+  /// Mark that this layer has changed and we need to upload changes to the
+  /// engine using [addToScene].
+  void markDirty() { _isDirty = true; }
+
+  // Whether any layer in the subtree of this layer is dirty ([_isDirty]).
+  bool _isSubtreeDirty;
+
+  ui.EngineLayer _engineLayer;
+
+  /// Traverse the layer tree and compute if any subtree has a dirty layer (by
+  /// calling [markDirty]).
+  void updateSubtreeDirtiness() {
+    _isSubtreeDirty = _isDirty;
+    if (this is ContainerLayer) {
+      final ContainerLayer container = this;
+      for (Layer child = container.firstChild; child != null; child = child.nextSibling) {
+        child.updateSubtreeDirtiness();
+        _isSubtreeDirty = _isSubtreeDirty || child._isSubtreeDirty;
+      }
+    }
+  }
```

Comment:
oh i see, it's not O(N^2) because you only run this once, before sending the layers up to the engine.

I recommend making this private, and having a single method that does all the logic of updating these bits, creating the SceneBuilder, calling addToScene, and eventually returning a Scene. That way the two places where you call this method would just call that method and there would be less danger of someone thinking they had to call this method sometimes. (That would also allow you to make addToScene protected.)